internal/pkg/metrics/incoming/ceilometer.go

```diff
@@ -1,73 +1,221 @@
+//TODO: Move this function to apputils
+func (c *CeilometerMetric) sanitize(data string) string {
+	sanitized := data
+	// parse only relevant data
+	sub := rexForOsloMessage.FindStringSubmatch(sanitized)
+	if len(sub) == 2 {
+		sanitized = rexForNestedQuote.ReplaceAllString(sub[1], `"`)
+	} else {
+		log.Printf("Failed to find oslo.message in given message: %s\n", data)
+	}
+	// avoid getting payload data wrapped in array
+	item := rexForPayload.FindStringSubmatch(sanitized)
+	if len(item) == 2 {
+		sanitized = rexForPayload.ReplaceAllString(sanitized, fmt.Sprintf(`"payload":%s`, item[1]))
+	}
+	return sanitized
+}
 
 //ParseInputJSON ...
-func (c *CeilometerMetric) ParseInputJSON(json string) ([]MetricDataFormat, error) {
-	return make([]MetricDataFormat, 0), nil
+func (c *CeilometerMetric) ParseInputJSON(data string) ([]MetricDataFormat, error) {
+	output := make([]MetricDataFormat, 0)
+	sanitized := c.sanitize(data)
+	// parse only relevant data
+	var json = jsoniter.ConfigCompatibleWithStandardLibrary
+	err := json.Unmarshal([]byte(sanitized), &c)
+	if err != nil {
+		return output, fmt.Errorf("error parsing json: %s", err)
+	}
+	c.SetNew(true)
+	c.SetData(c)
+	output = append(output, c)
+	return output, nil
 }
 
 //GetKey ...
 func (c *CeilometerMetric) GetKey() string {
-	return ""
+	return c.Publisher
 }
 
-//GetItemKey ...
+//GetItemKey returns name cache key analogically to CollectdMetric implementation
 func (c *CeilometerMetric) GetItemKey() string {
-	return ""
+	parts := []string{c.Plugin}
+	if c.Plugin != c.Type {
+		parts = append(parts, c.Type)
+	}
+	if c.PluginInstance != "" {
+		parts = append(parts, c.PluginInstance)
+	}
+	if c.TypeInstance != "" {
+		parts = append(parts, c.TypeInstance)
+	}
+	return strings.Join(parts, "_")
 }
 
-//ParseInputByte ...
+//ParseInputByte is not really used. It is here just to implement MetricDataFormat
+//TODO: Remove this method from here and also CollectdMetric
 func (c *CeilometerMetric) ParseInputByte(data []byte) error {
-	return nil
+	_, err := c.ParseInputJSON(string(data))
+	return err
 }
 
-//GetInterval ...
+//GetInterval returns hardcoded defaultCeilometerInterval, because Ceilometer metricDesc
+//does not contain interval information (are not periodically sent at all) and any reasonable
+//interval might be needed for expiry setting for Prometheus
+//TODO: Make this configurable
 func (c *CeilometerMetric) GetInterval() float64 {
-	return 0.0
+	return defaultCeilometerInterval
 }
 
 //SetNew ...
-func (c *CeilometerMetric) SetNew(new bool) {}
+func (c *CeilometerMetric) SetNew(new bool) {
+	c.new = new
+}
 
 //ISNew ...
 func (c *CeilometerMetric) ISNew() bool {
-	return true
+	return c.new
 }
 
 /*************************** tsdb.TSDB interface *****************************/
 
 //GetLabels ...
 func (c *CeilometerMetric) GetLabels() map[string]string {
-	return make(map[string]string)
+	labels := make(map[string]string)
+	if c.TypeInstance != "" {
```

Comment:
Just FYI, this re-labeling logic around plugin and type instances has been entirely removed from the collectd side of things in sg2. We are unclear on it's origin or purpose, but it made writing alarms and dashboards a bit more awkward rather than just using exactly what was in the collectd message. I have no opinion yet on whether it should exists here, since I haven't tried working with the resulting ceilometer metrics yet.

Reply:
Yeah the naming algorithm seems kinda strange to me to. I just wanted to be consistent witn collectd implementation. IMO if we want to remove it from Ceilometer, we should do this also for collectd. If it was removed from SG2, maybe we should do this as part of this PR prior to GA.

Reply:
I wouldn't bother with the renaming in this repo because once we productize SG.next we won't be using that code path any longer (even if the code remains). I don't want to accidentally cause a regression in the GA code if we need to release in the meantime.

I do think it's a good idea to do the refactoring here for the ceilometer metrics part to better align with SG.next since that's the final location for things to land ultimately.

Reply:
So we will have one type of metrics with old style names and the other types of metrics with new style? Fine by me, just don't like inconsistencies :).

Reply:
@paramite no, because you should converge to whatever the SG.next is doing. Once SG.next is ready, we will no longer use this code base to handle metrics delivery for collectd. We could just delete that path in the code base.

Alternatively align the naming conventions against the current collectd code base here to keep things consistent, then when porting the ceilometer metrics support into SG.next, change the naming scheme at that point to align.

I just feel like changing anything in the current collectd code path is really just a waste of time since we'll replace that path with SG.next when it available, and we're more likely to just create a regression for the current GA deployments if we need to release another version of this SG.

If we change collectd now, we'll need to also make sure that any dashboard changes align with this next release, and make sure customers upgrade the dashboard when upgrading SG. Seems like a lot of coordination in a product I'd like to avoid.

So either pick the naming used by SG.next to align and not need to change things later (probably my preference?) or align to what is here to keep all the metrics consistent (whether they be collectd or ceilometer) and then re-align the naming when porting to SG.next.

Not really sure which path is the correct one. Any other thoughts?